[Tools][git-webkit] Wrong identifiers for commits on some buildbot workers after the upgrade to 4.3.0

clopez opened the pull request:
#### d6777165fca3b2d3a9cddedc71e0d771c3805a78
<pre>
[Tools][git-webkit] Wrong identifiers for commits on some buildbot workers after the upgrade to 4.3.0
<a href="https://bugs.webkit.org/show_bug.cgi?id=299395">https://bugs.webkit.org/show_bug.cgi?id=299395</a>

Reviewed by Philippe Normand.

Since build.webkit.org buildbot server was upgraded the workers on the nighlty scheduler
are generating wrong commit identifiers, because git-webkit is not longer unable to tell
if the checked out commit belongs to the main branch or not.

This is caused because buildbot now does a checkout to the target commit instead of
resetting the main branch to it, so git ends in a detached state instead of on the branch.

Then git-webkit tries to see if the commit checked out belongs to main, but it fails
to do so because the bot does this to update and check the commit:
    git fetch -f --progress <a href="https://github.com/WebKit/WebKit.git">https://github.com/WebKit/WebKit.git</a> HEAD
    git checkout -f FETCH_HEAD

That means that git is not updating origin/remotes/main to FETCH_HEAD, so then git-webkit
thinks that FETCH_HEAD doesn&apos;t belong to main (or origin/remotes/main) because it is not
an ancestor commit (at least not until the reference for main gets updated to FETCH_HEAD)

To update the reference on origin/remotes/main one can force running &quot;git fetch&quot; or
&quot;git fetch origin&quot; (without passing an URL, but telling it to just fetch the origin).
But buildbot workers are not doing that.

This patch improves the tool git-webkit to look also at the commit in FETCH_HEAD when
trying to see if the commit belongs to the default branch. To achieve that it does:
 - Check if the merge-candidate commit in FETCH_HEAD is for the default branch and for
   the default repo URL.
 - Check if the reference for origin/remotes/${defaultbranch} is an ancestor of this FETCH_HEAD
   merge-candidate (fast-forward update).
 - Finally, update the reference for origin/remotes/{defaultbranch} to point to FETCH_HEAD

This patch also fixes a pre-existent logic problem in the method for mocking the behavior
of &apos;git merge-base --is-ancestor&apos; and fixes the affected unittests at checkoutroute.

* Tools/Scripts/libraries/reporelaypy/reporelaypy/tests/checkoutroute_unittest.py:
(CheckoutRouteUnittest.test_json_details_origin):
(CheckoutRouteUnittest):
(CheckoutRouteUnittest.test_json_strip_details_non_origin_commits):
(CheckoutRouteUnittest.test_json): Deleted.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git):
(Git._maybe_update_default_branch_ref_from_fetch_head):
(Git._is_on_default_branch):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
(TestGit._prepare_fetch_head_commit):

Canonical link: <a href="https://commits.webkit.org/301318@main">https://commits.webkit.org/301318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d18302055549cc2cba32a3e53f30c2e3a52c5e0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42412 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129321 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74804 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124576 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43131 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51005 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93252 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61921 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125652 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34390 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73893 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/122057 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33377 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28003 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-font-loading/idlharness.https.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72808 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104089 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28214 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132049 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37788 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101779 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/122121 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50022 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106061 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101659 "Found 4 new API test failures: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, WebKitGTK/TestWebKitNetworkSession:/webkit/WebKitNetworkSession/proxy, TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47025 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25196 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46415 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19666 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49502 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48969 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52321 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/834a7690e69f77dc2b61c5b9a574c75afcc97a4b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51261

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8f9a9a9a2b662dc63666509918d83617d1170230. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51261

clopez committed:
[Tools][git-webkit] Wrong identifiers for commits on some buildbot workers after the upgrade to 4.3.0
https://bugs.webkit.org/show_bug.cgi?id=299395

Reviewed by Philippe Normand.

Since build.webkit.org buildbot server was upgraded the workers on the nighlty scheduler
are generating wrong commit identifiers, because git-webkit is not longer unable to tell
if the checked out commit belongs to the main branch or not.

This is caused because buildbot now does a checkout to the target commit instead of
resetting the main branch to it, so git ends in a detached state instead of on the branch.

Then git-webkit tries to see if the commit checked out belongs to main, but it fails
to do so because the bot does this to update and check the commit:
    git fetch -f --progress https://github.com/WebKit/WebKit.git HEAD
    git checkout -f FETCH_HEAD

That means that git is not updating origin/remotes/main to FETCH_HEAD, so then git-webkit
thinks that FETCH_HEAD doesn't belong to main (or origin/remotes/main) because it is not
an ancestor commit (at least not until the reference for main gets updated to FETCH_HEAD)

To update the reference on origin/remotes/main one can force running "git fetch" or
"git fetch origin" (without passing an URL, but telling it to just fetch the origin).
But buildbot workers are not doing that.

This patch improves the tool git-webkit to look also at the commit in FETCH_HEAD when
trying to see if the commit belongs to the default branch. To achieve that it does:
 - Check if the merge-candidate commit in FETCH_HEAD is for the default branch and for
   the default repo URL.
 - Check if the reference for origin/remotes/${defaultbranch} is an ancestor of this FETCH_HEAD
   merge-candidate (fast-forward update).
 - Finally, update the reference for origin/remotes/{defaultbranch} to point to FETCH_HEAD

This patch also fixes a pre-existent logic problem in the method for mocking the behavior
of 'git merge-base --is-ancestor' and fixes the affected unittests at checkoutroute.

* Tools/Scripts/libraries/reporelaypy/reporelaypy/tests/checkoutroute_unittest.py:
(CheckoutRouteUnittest.test_json_details_origin):
(CheckoutRouteUnittest):
(CheckoutRouteUnittest.test_json_strip_details_non_origin_commits):
(CheckoutRouteUnittest.test_json): Deleted.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/local/git.py:
(Git):
(Git._maybe_update_default_branch_ref_from_fetch_head):
(Git._is_on_default_branch):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
(TestGit._prepare_fetch_head_commit):

Canonical link: https://commits.webkit.org/301318@main
